### PR TITLE
EVM refactor and simple UI

### PR DESCRIPTION
**Pull request by feliam**
Goal/s 
Be able to do what seth.report used to do in batch in a more controllable way.
Be able to browse states in the running and terminated lists.
Be able to browse the transactions on each state easely
Be able to parse the calldata and return_data using the ABI from contracts that came from a solidity sourcecode.

What's done?
The api to query states and "blockchain" world is now usable (as in existent).

python evm/axamples/solidse.py sourcecode.sol

refunds, reverts, selfdestructed contracts need testing

### CODE REVIEW

**Comment by mossberg** [comment]
I get this error running reentrancy_symbolic.py

```
/tmp/tmpZzD0p4:40:13: Warning: Return value of low-level calls not used.
            vulnerable_contract.call(reentry_attack_string);
            ^---------------------------------------------^
Traceback (most recent call last):
  File "reentrancy_symbolic.py", line 82, in <module>
    exploit_account = seth.solidity_create_contract(exploit_source_code, owner=attacker_account)
  File "/mnt/hgfs/code/manticore/manticore/seth.py", line 508, in solidity_create_contract
    self.metadata[address] = SolidityMetadata(name, source_code, init_bytecode, metadata, metadata_runtime, hashes, abi)
  File "/mnt/hgfs/code/manticore/manticore/seth.py", line 38, in __init__
    self.abi = dict( [(item['name'], item) for item in abi ])
KeyError: 'name'

```

**Comment by mossberg** [comment]
i get this error when running solidse.py on [this](https://trailofbits.slack.com/files/U1P0JM3D2/F8ABL5TG9/superwallet.sol)

```
[+] Creating a contract account 0x1bfa530d5d685155e98cd7d9dd23f7b6a801cfefL
[+] Creating a attacker account 1216709658712221897132306291703014113940054544883
2017-12-06 18:59:59,412: [7308] m.c.executor:ERROR: Exception: global name 'data' is not defined
Traceback (most recent call last):
  File "/mnt/hgfs/code/manticore/manticore/core/executor.py", line 463, in run
    if not current_state.execute():
  File "/mnt/hgfs/code/manticore/manticore/core/state.py", line 126, in execute
    result = self._platform.execute()
  File "/mnt/hgfs/code/manticore/manticore/platforms/evm.py", line 2080, in execute
    self.STOP()
  File "/mnt/hgfs/code/manticore/manticore/platforms/evm.py", line 2274, in STOP
    tx.return_data=data
NameError: global name 'data' is not defined

```

File: manticore/seth.py [hunk 1]
```diff
@@ -9,6 +9,60 @@
...
+        return '('+','.join(x['type'] for x in abi['outputs']) +')'
+
+    def get_func_name(self, hsh):
+        signature = self.signatures.get(hsh,'{fallback}()')
+        return signature.split('(')[0]
+        
+    def get_func_signature(self, hsh):
+        return self.signatures.get(hsh, '{fallback}()')
+
+    def __redduce__(self):
```
Comment (mossberg) [inline manticore/seth.py]: typo

**Comment by mossberg** [comment]
I got this with minimal.py

```
[I] mark ubuntu /m/h/c/m/e/evm (dev-evm-refactor) ❯ py minimal.py
[+] Creating a user account
[+] Creating a contract account
/tmp/tmpcuG6eh:6:5: Warning: No visibility specified. Defaulting to "public".
    function() payable {
    ^
Spanning multiple lines.
[+] Now the symbolic values
Traceback (most recent call last):
  File "minimal.py", line 40, in <module>
    print "[+] There are %d reverted states now"% len(seth.final_state_ids)
AttributeError: 'ManticoreEVM' object has no attribute 'final_state_ids'
```

**Comment by mossberg** [comment]
and with coverage.py

```
[I] mark ubuntu /m/h/c/m/e/evm (dev-evm-refactor) ❯ py coverage.py
/tmp/tmprDX5ez:41:5: Warning: No visibility specified. Defaulting to "public".
    function balanceOf(address user)
    ^
Spanning multiple lines.
/tmp/tmprDX5ez:41:5: Warning: Function state mutability can be restricted to view
    function balanceOf(address user)
    ^
Spanning multiple lines.
Traceback (most recent call last):
  File "coverage.py", line 10, in <module>
    bytecode = seth.compile(source_code)
  File "/mnt/hgfs/code/manticore/manticore/seth.py", line 355, in compile
    name, source_code, bytecode, srcmap, srcmap_runtime, hashes = ManticoreEVM._compile(source_code)
ValueError: too many values to unpack
```

**Comment by mossberg** [comment]
reentrancy_concrete.py

```
[+] Let attacker extract all  using exploit
[+] Let attacker destroy the exploit andprofit
2017-12-07 11:59:37,837: [28194] m.c.executor:ERROR: Exception: 'EVMWorld' object has no attribute 'suicide'
Traceback (most recent call last):
  File "/mnt/hgfs/code/manticore/manticore/core/executor.py", line 463, in run
    if not current_state.execute():
  File "/mnt/hgfs/code/manticore/manticore/core/state.py", line 126, in execute
    result = self._platform.execute()
  File "/mnt/hgfs/code/manticore/manticore/platforms/evm.py", line 2086, in execute
    self.SELFDESTRUCT(ex.to)
  File "/mnt/hgfs/code/manticore/manticore/platforms/evm.py", line 2327, in SELFDESTRUCT
    self.suicide.add(address)
AttributeError: 'EVMWorld' object has no attribute 'suicide'

Traceback (most recent call last):
  File "reentrancy_concrete.py", line 122, in <module>
    exploit_account.get_money()
  File "/mnt/hgfs/code/manticore/manticore/seth.py", line 282, in f
    data=tx_data
  File "/mnt/hgfs/code/manticore/manticore/seth.py", line 571, in transaction
    return  self.run(procs=self._config_procs)
  File "/mnt/hgfs/code/manticore/manticore/seth.py", line 594, in run
    self._initial_state = self._executor._workspace.load_state(context['_saved_states'][0], delete=True)
  File "/mnt/hgfs/code/manticore/manticore/core/workspace.py", line 382, in load_state
    return self._store.load_state('{}{:08x}{}'.format(self._prefix, state_id, self._suffix), delete=delete)
  File "/mnt/hgfs/code/manticore/manticore/core/workspace.py", line 163, in load_state
    state = self._serializer.deserialize(f)
  File "/mnt/hgfs/code/manticore/manticore/core/workspace.py", line 51, in deserialize
    return cPickle.load(f)
TypeError: ('__init__() takes exactly 2 arguments (1 given)', <class 'manticore.platforms.evm.SelfDestruct'>, ())
```

File: manticore/core/executor.py [hunk 2]
```diff
@@ -427,23 +427,26 @@ def run(self):
...
-                                current_state = self._workspace.load_state(current_state_id)
-                                self.forward_events_from(current_state, True)
-                                self._publish('did_load_state', current_state, current_state_id)
-                                logger.info("load state %r", current_state_id)
-                            #notify siblings we have a state to play with
-                            self._notify_start_run()
+                try:  # handle fatal errors: exceptions in Manticore
+                    try:  # handle external (e.g. solver) errors, and executor control exceptions
+                        #select a suitable state to analyze
+                        if current_state is None:
```
Comment (mossberg) [inline manticore/core/executor.py]: what exceptions were you experiencing in this code? Concretize/TerminateState/SolverException?

Comment (feliam) [reply manticore/core/executor.py]: Some failing exception from the event callback. Any exception would to break it.

Comment (mossberg) [reply manticore/core/executor.py]: wouldn't those still be caught by the bottom except block for Exceptions, AssertionError? moving the try up here doesn't do anything unless the exception from this block is a Concretize/TerminateState/SolverException?

File: manticore/manticore.py [hunk 3]
```diff
@@ -394,7 +394,7 @@ def wrapper(*args, **kwargs):
 
     @property
     def running(self):
-        return self._executor._running.value
```
Comment (mossberg) [inline manticore/manticore.py]: `_executor.running` is a SyncManager.Value(). don't we want to return the `.value`, not the Value() directory?

Comment (mossberg) [reply manticore/manticore.py]: whoops ignore

File: manticore/core/executor.py (hunk 2 above)
Comment (feliam) [reply manticore/core/executor.py]: The new refactor _hopefully_ maintain the  
`
#{self._notify_start_run()} = #{self._notify_stop_run()}
`
invariant, even if exceptions are raised from the callbacks.

File: manticore/core/smtlib/solver.py [hunk 4]
```diff
@@ -461,7 +468,7 @@ def get_value(self, constraints, expression):
                 var = temp_cs.new_bool()
             elif isinstance(expression, BitVec):
                 var = temp_cs.new_bitvec(expression.size)
-            elif isinstance(expression, Array):
+            elif isinstance(expression, (Array, ArrayProxy)):
```
Comment (yan) [inline manticore/core/smtlib/solver.py]: Is the check for `ArrayProxy` necessary, as its an `Array` subclass?

Comment (feliam) [reply manticore/core/smtlib/solver.py]: It should not be necessary! Let's remove it and see.

File: manticore/platforms/evm.py [hunk 5]
```diff
@@ -1948,6 +1997,28 @@ def set_code(self, address, data):
         self.storage[address]['code'] = data
 
 
+    def log(self, address, topic, data):
+        self.logs.append((address, data, topics))
+        logger.info('LOG %r %r', memlog, topics)
+
+
+    '''
```
Comment (yan) [inline manticore/platforms/evm.py]: Remove commented out block?

File: manticore/platforms/evm.py [hunk 6]
```diff
@@ -1970,17 +2041,18 @@ def _pop_vm(self, rollback=False):
...
                 self._global_storage = vm.global_storage
 
-            self._deleted_address = self._deleted_address.union(vm.suicide)
-            self._logs += vm.logs
-            if not self.depth:
-                for address in self._deleted_address:
-                    del self.storage[address]
+                self._deleted_address = self._deleted_address.union(vm.suicides)
+                self._logs += vm.logs
+                if not self.depth:
```
Comment (mossberg) [inline manticore/platforms/evm.py]: this is redundant

File: manticore/platforms/evm.py [hunk 7]
```diff
@@ -2165,8 +2246,10 @@ def CALL(self, gas, to, value, data):
     def RETURN(self, data):
         prev_vm = self._pop_vm() #current VM changed!
         if self.depth == 0:
-            self.last_return=data
-            self.last_pc = prev_vm.pc
+            tx = self._transactions[-1]
+            tx.return_data=data
+            #tx.last_pc = prev_vm.pc
```
Comment (mossberg) [inline manticore/platforms/evm.py]: why don't we record the prev_vm.pc in the new version?

File: manticore/platforms/evm.py [hunk 8]
```diff
@@ -2236,7 +2327,9 @@ def SELFDESTRUCT(self, recipient):
         self.suicide.add(address)
         prev_vm = self._pop_vm(rollback=False)
         if self.depth == 0:
-            self.last_pc = prev_vm.pc
+            tx = self._transactions[-1]
+            tx.last_pc = prev_vm.pc
+            tx.result_type = 'SELFDESTRUCT'
```
Comment (mossberg) [inline manticore/platforms/evm.py]: tx.result_type doesn't exist?

File: manticore/platforms/evm.py (hunk 7 above)
Comment (feliam) [reply manticore/platforms/evm.py]: It's not really part of a transaction. So there is not room for it in tx. 
and... it was not used?. I could not answer: Why do we used to save it ?

File: manticore/platforms/evm.py (hunk 8 above)
Comment (feliam) [reply manticore/platforms/evm.py]: No.  It's just .result now. 
That fix sound familiar. .. checking.

File: manticore/seth.py [hunk 9]
```diff
@@ -441,12 +631,9 @@ def load(self, state_id=None):
             if len(self.running_state_ids) == 1:  
                 #Get the ID of the single running state              
                 state_id = self.running_state_ids[0]
-                if state_id != -1:
-                    #if there is a single running state with id != 1. We consider it is a new initial_state
-                    assert self.initial_state is None
-                    state = self.initial_state = self._executor._workspace.load_state(state_id, delete=True)
             else:
-                raise Exception("More than one state running. Do not know which to choose.")
+                raise Exception("More than one state running. Do not know which one to choose.")
```
Comment (mossberg) [inline manticore/seth.py]: for now, does it matter which state to choose? just choose the first?

Comment (feliam) [reply manticore/seth.py]: None, means "Give me the one state". It should fail if multiple states are enqueuedueue. 
I'll change the txt. to  "More than one state running."

File: manticore/seth.py [hunk 10]
```diff
@@ -528,8 +725,21 @@ def _did_read_code(self, state, offset, size):
...
-        ''' Prints a small report on state id '''
+
+    def get_metadata(self, address):
+        ''' Gets the solidity metadata for address.
+            This is available only if address is a contract created from solidity
+        '''
+        try:
+            md = self.metadata[address]
+        except:
+            md = SolidityMetadata(None, None, None, None, None, None)
```
Comment (mossberg) [inline manticore/seth.py]: imo we should return `None` instead of a fake SolidityMetadata

**Pull request merged**